File: elixir/lib/api_requests.ex
defmodule ApiRequests do
  @moduledoc """
  API Request Performance Test for Elixir.
  Uses Task.async_stream for efficient concurrent HTTP requests.
  Collects comprehensive metrics including response times, throughput, and percentiles.
  """

  @doc """
  Make a single HTTP request and return timing data.
  """
  def make_request(url, request_id, start_time) do
    request_start = :os.system_time(:millisecond)
    
    result = case HTTPoison.get(url, [], recv_timeout: 10_000, timeout: 10_000) do
      {:ok, %HTTPoison.Response{status_code: 200, body: _body}} ->
        request_end = :os.system_time(:millisecond)
        %{
          id: request_id,
          success: true,
          response_time_ms: request_end - request_start,
          timestamp: (request_end - start_time) / 1000.0
        }
      
      {:ok, %HTTPoison.Response{status_code: status}} ->
        request_end = :os.system_time(:millisecond)
        %{
          id: request_id,
          success: false,
          response_time_ms: request_end - request_start,
          timestamp: (request_end - start_time) / 1000.0,
          error: "HTTP #{status}"
        }
      
      {:error, reason} ->
        request_end = :os.system_time(:millisecond)
        %{
          id: request_id,
          success: false,
          response_time_ms: request_end - request_start,
          timestamp: (request_end - start_time) / 1000.0,
          error: inspect(reason)
        }
    end
    
    result
  end

  @doc """
  Calculate percentile from sorted list.
  """
  def percentile(sorted_list, p) do
    index = round(length(sorted_list) * p) - 1
    index = max(0, min(index, length(sorted_list) - 1))
    Enum.at(sorted_list, index)
  end

  @doc """
  Run API request benchmark with 10,000 concurrent requests.
  """
  def run_benchmark do
    url = "https://jsonplaceholder.typicode.com/posts/1"
    num_requests = 1_000
    
    IO.puts("Starting benchmark: #{num_requests} requests to #{url}")
    
    # Start HTTPoison application
    HTTPoison.start()
    
    # Measure total time
    start_time = :os.system_time(:millisecond)
    
    # Make concurrent requests using Task.async_stream
    results = 0..(num_requests - 1)
              |> Task.async_stream(
                fn i -> make_request(url, i, start_time) end,
                max_concurrency: 100,
                timeout: 15_000
              )
              |> Enum.map(fn {:ok, result} -> result end)
    
    end_time = :os.system_time(:millisecond)
    total_time = (end_time - start_time) / 1000.0  # Convert to seconds
    
    # Analyze results
    successful_results = Enum.filter(results, fn r -> r.success end)
    failed_results = Enum.filter(results, fn r -> !r.success end)
    
    successful_count = length(successful_results)
    failed_count = length(failed_results)
    
    # Extract and sort response times
    response_times = Enum.map(successful_results, fn r -> r.response_time_ms * 1.0 end)
    sorted_times = Enum.sort(response_times)
    
    # Calculate metrics
    metrics = if length(response_times) > 0 do
      %{
        language: "elixir",
        total_requests: num_requests,
        successful_requests: successful_count,
        failed_requests: failed_count,
        total_time_seconds: Float.round(total_time, 2),
        requests_per_second: Float.round(num_requests / total_time, 2),
        response_times: %{
          min_ms: Float.round(Enum.min(response_times), 2),
          max_ms: Float.round(Enum.max(response_times), 2),
          average_ms: Float.round(Enum.sum(response_times) / length(response_times), 2),
          median_ms: Float.round(percentile(sorted_times, 0.50), 2),
          p95_ms: Float.round(percentile(sorted_times, 0.95), 2),
          p99_ms: Float.round(percentile(sorted_times, 0.99), 2)
        },
        timeseries: Enum.map(successful_results, fn r ->
          %{
            timestamp: Float.round(r.timestamp, 3),
            response_time_ms: Float.round(r.response_time_ms * 1.0, 2)
          }
        end)
      }
    else
      %{
        language: "elixir",
        total_requests: num_requests,
        successful_requests: 0,
        failed_requests: failed_count,
        total_time_seconds: Float.round(total_time, 2),
        requests_per_second: 0,
        response_times: %{},
        timeseries: []
      }
    end
    
    # Save results to JSON file
    output_file = "api_results.json"
    json_data = Jason.encode!(metrics, pretty: true)
    File.write!(output_file, json_data)
    
    IO.puts("\nResults saved to #{output_file}")
    
    metrics
  end

  @doc """
  Print formatted results to console.
  """
  def print_results(metrics) do
    IO.puts("\n#{"=" |> String.duplicate(60)}")
    IO.puts("Language: #{String.upcase(metrics.language)}")
    IO.puts("#{"=" |> String.duplicate(60)}")
    IO.puts("Total Requests: #{metrics.total_requests}")
    IO.puts("Successful: #{metrics.successful_requests}")
    IO.puts("Failed: #{metrics.failed_requests}")
    IO.puts("Total Time: #{metrics.total_time_seconds}s")
    IO.puts("Requests/sec: #{metrics.requests_per_second}")
    
    if map_size(metrics.response_times) > 0 do
      rt = metrics.response_times
      IO.puts("\nResponse Times (ms):")
      IO.puts("  Min: #{rt.min_ms}")
      IO.puts("  Max: #{rt.max_ms}")
      IO.puts("  Avg: #{rt.average_ms}")
      IO.puts("  Median: #{rt.median_ms}")
      IO.puts("  P95: #{rt.p95_ms}")
      IO.puts("  P99: #{rt.p99_ms}")
    end
    
    IO.puts("#{"=" |> String.duplicate(60)}\n")
  end
end

# Run if this is the main file
if System.argv() |> Enum.member?("--run") do
  metrics = ApiRequests.run_benchmark()
  ApiRequests.print_results(metrics)
end

File: scala/src/main/scala/ApiRequests.scala
import sttp.client3._
import scala.concurrent.{Await, Future, ExecutionContext}
import scala.concurrent.duration._
import java.io.PrintWriter
import scala.util.{Try, Success, Failure}

/**
 * API Request Performance Test for Scala.
 * Uses sttp client with Futures for efficient concurrent HTTP requests.
 * Collects comprehensive metrics including response times, throughput, and percentiles.
 */
object ApiRequests {
  implicit val ec: ExecutionContext = ExecutionContext.global

  case class RequestResult(
    id: Int,
    success: Boolean,
    responseTimeMs: Double,
    timestamp: Double,
    error: Option[String] = None
  )

  case class ResponseTimes(
    min_ms: Double,
    max_ms: Double,
    average_ms: Double,
    median_ms: Double,
    p95_ms: Double,
    p99_ms: Double
  )

  case class TimeseriesPoint(
    timestamp: Double,
    response_time_ms: Double
  )

  case class ApiMetrics(
    language: String,
    total_requests: Int,
    successful_requests: Int,
    failed_requests: Int,
    total_time_seconds: Double,
    requests_per_second: Double,
    response_times: Option[ResponseTimes],
    timeseries: List[TimeseriesPoint]
  )

  /**
   * Make a single HTTP request asynchronously.
   */
  def makeRequest(backend: SttpBackend[Identity, Any], url: String, requestId: Int, startTime: Long): RequestResult = {
    val requestStart = System.currentTimeMillis()
    
    try {
      val request = basicRequest
        .get(uri"$url")
        .readTimeout(10.seconds)
      
      val response = request.send(backend)
      val requestEnd = System.currentTimeMillis()
      
      RequestResult(
        id = requestId,
        success = response.code.code == 200,
        responseTimeMs = (requestEnd - requestStart).toDouble,
        timestamp = (requestEnd - startTime) / 1000.0
      )
    } catch {
      case e: Exception =>
        val requestEnd = System.currentTimeMillis()
        RequestResult(
          id = requestId,
          success = false,
          responseTimeMs = (requestEnd - requestStart).toDouble,
          timestamp = (requestEnd - startTime) / 1000.0,
          error = Some(e.getMessage)
        )
    }
  }

  /**
   * Calculate percentile from sorted list.
   */
  def percentile(sortedList: List[Double], p: Double): Double = {
    val index = math.max(0, math.min((sortedList.length * p).toInt - 1, sortedList.length - 1))
    sortedList(index)
  }

  /**
   * Run API request benchmark with 10,000 concurrent requests.
   */
  def runBenchmark(): ApiMetrics = {
    val url = "https://jsonplaceholder.typicode.com/posts/1"
    val numRequests = 10000

    println(s"Starting benchmark: $numRequests requests to $url")

    val backend = HttpURLConnectionBackend()
    val startTime = System.currentTimeMillis()

    // Make concurrent requests using parallel collections
    import scala.collection.parallel.CollectionConverters._
    val results = (0 until numRequests).par.map { i =>
      makeRequest(backend, url, i, startTime)
    }.toList

    val endTime = System.currentTimeMillis()
    val totalTime = (endTime - startTime) / 1000.0 // Convert to seconds

    backend.close()

    // Analyze results
    val successfulResults = results.filter(_.success)
    val failedResults = results.filter(!_.success)

    val successfulCount = successfulResults.length
    val failedCount = failedResults.length

    // Extract and sort response times
    val responseTimes = successfulResults.map(_.responseTimeMs)
    val sortedTimes = responseTimes.sorted

    // Create metrics
    val metrics = if (responseTimes.nonEmpty) {
      ApiMetrics(
        language = "scala",
        total_requests = numRequests,
        successful_requests = successfulCount,
        failed_requests = failedCount,
        total_time_seconds = BigDecimal(totalTime).setScale(2, BigDecimal.RoundingMode.HALF_UP).toDouble,
        requests_per_second = BigDecimal(numRequests / totalTime).setScale(2, BigDecimal.RoundingMode.HALF_UP).toDouble,
        response_times = Some(ResponseTimes(
          min_ms = BigDecimal(responseTimes.min).setScale(2, BigDecimal.RoundingMode.HALF_UP).toDouble,
          max_ms = BigDecimal(responseTimes.max).setScale(2, BigDecimal.RoundingMode.HALF_UP).toDouble,
          average_ms = BigDecimal(responseTimes.sum / responseTimes.length).setScale(2, BigDecimal.RoundingMode.HALF_UP).toDouble,
          median_ms = BigDecimal(percentile(sortedTimes, 0.50)).setScale(2, BigDecimal.RoundingMode.HALF_UP).toDouble,
          p95_ms = BigDecimal(percentile(sortedTimes, 0.95)).setScale(2, BigDecimal.RoundingMode.HALF_UP).toDouble,
          p99_ms = BigDecimal(percentile(sortedTimes, 0.99)).setScale(2, BigDecimal.RoundingMode.HALF_UP).toDouble
        )),
        timeseries = successfulResults.map(r =>
          TimeseriesPoint(
            timestamp = BigDecimal(r.timestamp).setScale(3, BigDecimal.RoundingMode.HALF_UP).toDouble,
            response_time_ms = BigDecimal(r.responseTimeMs).setScale(2, BigDecimal.RoundingMode.HALF_UP).toDouble
          )
        )
      )
    } else {
      ApiMetrics(
        language = "scala",
        total_requests = numRequests,
        successful_requests = 0,
        failed_requests = failedCount,
        total_time_seconds = BigDecimal(totalTime).setScale(2, BigDecimal.RoundingMode.HALF_UP).toDouble,
        requests_per_second = 0.0,
        response_times = None,
        timeseries = List.empty
      )
    }

    // Save results to JSON file
    val outputFile = "api_results.json"
    saveToJson(metrics, outputFile)

    println(s"\nResults saved to $outputFile")

    metrics
  }

  /**
   * Save metrics to JSON file.
   */
  def saveToJson(metrics: ApiMetrics, filename: String): Unit = {
    import ujson._

    val json = Obj(
      "language" -> metrics.language,
      "total_requests" -> metrics.total_requests,
      "successful_requests" -> metrics.successful_requests,
      "failed_requests" -> metrics.failed_requests,
      "total_time_seconds" -> metrics.total_time_seconds,
      "requests_per_second" -> metrics.requests_per_second,
      "response_times" -> (metrics.response_times match {
        case Some(rt) => Obj(
          "min_ms" -> rt.min_ms,
          "max_ms" -> rt.max_ms,
          "average_ms" -> rt.average_ms,
          "median_ms" -> rt.median_ms,
          "p95_ms" -> rt.p95_ms,
          "p99_ms" -> rt.p99_ms
        )
        case None => Obj()
      }),
      "timeseries" -> Arr(
        metrics.timeseries.map(ts =>
          Obj(
            "timestamp" -> ts.timestamp,
            "response_time_ms" -> ts.response_time_ms
          )
        ): _*
      )
    )

    val writer = new PrintWriter(filename)
    writer.write(ujson.write(json, indent = 2))
    writer.close()
  }

  /**
   * Print formatted results to console.
   */
  def printResults(metrics: ApiMetrics): Unit = {
    println("\n" + "=" * 60)
    println(s"Language: ${metrics.language.toUpperCase}")
    println("=" * 60)
    println(s"Total Requests: ${metrics.total_requests}")
    println(s"Successful: ${metrics.successful_requests}")
    println(s"Failed: ${metrics.failed_requests}")
    println(f"Total Time: ${metrics.total_time_seconds}%.2fs")
    println(f"Requests/sec: ${metrics.requests_per_second}%.2f")

    metrics.response_times.foreach { rt =>
      println("\nResponse Times (ms):")
      println(f"  Min: ${rt.min_ms}%.2f")
      println(f"  Max: ${rt.max_ms}%.2f")
      println(f"  Avg: ${rt.average_ms}%.2f")
      println(f"  Median: ${rt.median_ms}%.2f")
      println(f"  P95: ${rt.p95_ms}%.2f")
      println(f"  P99: ${rt.p99_ms}%.2f")
    }

    println("=" * 60 + "\n")
  }

  def main(args: Array[String]): Unit = {
    val metrics = runBenchmark()
    printResults(metrics)
  }
}
